samples/99_zil_interpreter/app/parser.rb

```diff
@@ -0,0 +1,375 @@
+# Parser class uses the scanner to load from file or string buffer. Results are loaded into expressions.
+#
+# Usage:
+#   source = "<ZIL ()>"
+#   parser.parse_string(args, source)
+#   expr0 = parser.expressions[0]
+#
+#   OR
+#
+#   parser.parse_file(args, '/data/zil/zork1/', 'zork1.zil')
+#   expr0 = parser.expressions[0]
+#
+class Parser
+  attr_reader :expressions
+
+  def initialize
+    @expressions = []
+  end
+
+  def parse_file(args, work_dir, file_name)
+    @work_dir = work_dir # not sure if we need work_dir, the idea was to follow INSERT-FILEs inside the parser/scanner
+
+    file = args.gtk.read_file(@work_dir + file_name)
+    log "Opening '" + file_name + "' ..."
+
+    buffer = ""
+    file.each_line.each { |line|
+      line.each_char { |c| buffer << c }
+    }
+
+    log "Scanning '" + file_name + "' ..."
+    scanner = Scanner.new(buffer)
+
+    i = 0
+    while scanner.scan_to('<;"%') # find next form, string, or comment
+      expr = scanner.read_expression
+      @expressions << expr
+      log "[#{file_name}:#{i}] " + expr.to_s
+      i += 1
+    end
+
+    log "Scan completed."
+  end
+
+  def parse_string(args, buffer)
+    log "Scanning buffer ..."
+    scanner = Scanner.new(buffer)
+
+    i = 0
+    while scanner.scan_to('<;"%') # find next form, string, or comment
+      expr = scanner.read_expression
+      @expressions << expr
+      log "[BUFFER:#{i}] " + expr.to_s
+      i += 1
+    end
+
+    log "Scan completed."
+  end
+end
+
+# This class can probably be reworked into Parser:
+#
+# Usage:
+#   scanner = Scanner.new(buffer)
+#   while scanner.scan_to('<;"%') # find next form, string, or comment
+#     expr = scanner.read_expression
+#   end
+#
+class Scanner
+  ATOM_CHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-+*/?=.,!$"
+  FIX_CHARS = '0123456789*#' # * = octal, #2 = binary
+  WHITESPACE_CHARS = " \r\n\t"
+
+  def initialize(buffer)
+    @buffer = buffer
+    @expr_depth = 0
+    @list_depth = 0
+    reset
+  end
+
+  def reset
+    @buf_index = -1
+  end
+
+  def scan_to(scan_chars)
+    found = false
+
+    while @buf_index < @buffer.length - 1
+      @buf_index += 1
+      buf_char = @buffer[@buf_index]
+
+      if scan_chars.include?(buf_char)
+        found = true
+        break
+      end
+
+    end
+
+    found
+  end
+
+  def skip_whitespace
+    done = false
+
+    until done

```

Comment:
I believe this can be reduced into a one-liner 
`@buf_index += 1 until WHITESPACE_CHARS.include?(@buffer[@buf_index]) || @buf_index >= @buffer.length`

Reply:
I reworked this method.  You may still have comments on it.